File: 7-handy-haversacks/main.exs
defmodule Main do
  def lines() do
    File.read!( "./input.txt" )
      |> String.trim() # Removes the last line
      |> String.split( "\n" ) # List of lines
  end

  def number_of_bags( bags ) do
    bags
      |> Enum.map( &( &1 |> hd() |> String.split() ) ) # list of list of words
      |> Enum.map( &[ String.to_integer(hd( &1 )), Enum.join(tl( &1 ), " ") ] ) # Tuple of bag ammount and bag type
  end

  def bags( lines ) do
    lines
      |> Enum.map( &(Regex.scan( ~r/(^\w+\s\w+|\d\s\w+\s\w+)/, &1, capture: :all_but_first )) ) # list of bags and ammount strings
      |> Enum.map( &[ hd( &1 ) |> hd(), number_of_bags( tl(&1) ) ] ) # List of bag type and containing bags
  end

  def find_parent_bags( searches, bags, searched, found) do
    case bags do
      [ [ type , contents ] = bag | bs ] ->
        case searches do
          [ search | _ ] -> 
            matched = Enum.any?( contents, fn([ _ , t ]) -> t === search end )

            tracked = Enum.member?( found, type )


            cond do
              matched and not tracked ->
                find_parent_bags( searches ++ [ type ], bs, [ bag | searched ], [ type | found ] )
              true -> 
                find_parent_bags( searches, bs, [ bag | searched ], found )
            end
          _ -> found
        end
      _ ->
        case searches do
          [ _ | ss ] -> find_parent_bags( ss, searched, [], found ) 
          _ -> found
        end
    end
  end

  def find_containing_bags( search, bags, searched ) do
    case bags do
      [ bag | bs ] ->
        case bag do
          [ type, contents ] when type === search and length( contents ) === 0 -> 0
          [ type, contents ] when type === search -> 
            contents
              |> Enum.map( fn([ num, t ]) ->
                num + (num * find_containing_bags( t, searched ++ bs, [] )) end )
              |> Enum.sum()
              #|> (&(IO.inspect(&1) && &1)).()
          _ ->
            find_containing_bags( search, bs, [ bag | searched ] )
        end
      _ -> 0
    end
  end

  def run() do
    lines()
      |> bags
      #|> (&(IO.inspect(&1) && &1)).()
      #|> (&find_parent_bags( [ "shiny gold" ], &1, [], [] )).() # Part one
      #|> (&IO.inspect(Kernel.length( &1 ))).()
      |> (&find_containing_bags( "shiny gold", &1, [] )).() # Part two
      |> (&IO.inspect/1).()
  end
end

Main.run()

:c.q

File: 7-handy-haversacks/main.js
/*
const lines = (await Deno.readTextFile("./input.test.txt"))
  .split("\n")
  .filter(Boolean);
*/

const number_of_bags = (bags) => {
  const [num, ...type] = bags.split(" ");

  return [num, type.join(" ")];
};

const bags = lines => lines
  .map((line) => line.match(/(^|[0-9]\s)(.\w+)(\s\w+)/g))
  .map(([bag, ...contains]) => [bag, contains.map(number_of_bags)]);

const find_parent_bags = (
  [search, ...searches],
  [bag, ...bags],
  searched,
  found = []
) => {
  if (!bag)
    return !search ? found : find_parent_bags(searches, searched, [], found);

  const [type, contents] = bag;

  const in_found = found.includes(type) ? [] : [type];

  const args = contents.some(([_, t]) => t === search)
    ? [
        [search, ...searches, ...in_found],
        bags,
        [bag, ...searched],
        [...in_found, ...found],
      ]
    : [[search, ...searches], bags, [bag, ...searched], found];

  return find_parent_bags(...args);
};

console.log(find_parent_bags(["shiny gold"], bags( lines ), []));

/*
function* find_parent_bags(
  [search, ...searches],
  [bag, ...bags],
  searched,
  found = []
) {
  if (!bag && search)
    return yield* find_parent_bags(searches, searched, [], found)

  const [type, contents] = bag;

  const in_found = found.includes(type) ? [] : [type];

  const contains_bag = contents.some(([_, t]) => t === search)
  
  if ( contains_bag )
    yield type

  const args = contains_bag
    ? [
        [search, ...searches, ...in_found],
        bags,
        [bag, ...searched],
        [...in_found, ...found],
      ]
    : [[search, ...searches], bags, [bag, ...searched], found];

  yield* find_parent_bags(...args);
}

for ( const i of find_parent_bags(["shiny gold"], bags, []) )
  console.log( i )
*/
